File: server/app/simulation/race_engine/course_data.py
"""Port of CourseData.ts."""

from __future__ import annotations

import json
import math
from pathlib import Path
from typing import NamedTuple

# Phase = 0 | 1 | 2 | 3


class Surface:
    TURF = 1
    DIRT = 2


class DistanceType:
    SHORT = 1
    MILE = 2
    MID = 3
    LONG = 4


class Orientation:
    CLOCKWISE = 1
    COUNTERCLOCKWISE = 2
    UNUSED = 3
    NO_TURNS = 4


class ThresholdStat:
    SPEED = 1
    STAMINA = 2
    POWER = 3
    GUTS = 4
    INT = 5


class Corner(NamedTuple):
    start: float
    length: float


class Straight(NamedTuple):
    start: float
    end: float
    front_type: int


class Slope(NamedTuple):
    start: float
    length: float
    slope: float


class CourseData:
    __slots__ = (
        "race_track_id", "distance", "distance_type", "surface", "turn",
        "course_set_status", "corners", "straights", "slopes", "max_lane",
    )

    def __init__(self, race_track_id, distance, distance_type, surface, turn,
                 course_set_status, corners, straights, slopes, max_lane):
        self.race_track_id = race_track_id
        self.distance = distance
        self.distance_type = distance_type
        self.surface = surface
        self.turn = turn
        self.course_set_status = course_set_status
        self.corners = corners
        self.straights = straights
        self.slopes = slopes
        self.max_lane = max_lane


_DATA_PATH = Path(__file__).resolve().parent / "data" / "course_data.json"
_raw_courses: dict | None = None
_course_cache: dict[int, CourseData] = {}


def _raw() -> dict:
    global _raw_courses
    if _raw_courses is None:
        _raw_courses = json.loads(_DATA_PATH.read_text(encoding="utf-8"))
    return _raw_courses


def is_sorted_by_start(items) -> bool:
    last = -1
    for item in items:
        if item.start <= last:
            return False
        last = item.start
    return True


def phase_start(distance: float, phase: int) -> float:
    if phase == 0:
        return 0
    if phase == 1:
        return distance * 1 / 6
    if phase == 2:
        return distance * 2 / 3
    if phase == 3:
        return distance * 5 / 6
    raise ValueError(phase)


def phase_end(distance: float, phase: int) -> float:
    if phase == 0:
        return distance * 1 / 6
    if phase == 1:
        return distance * 2 / 3
    if phase == 2:
        return distance * 5 / 6
    if phase == 3:
        return distance
    raise ValueError(phase)


def course_speed_modifier(course: CourseData, stats) -> float:
    """`stats` needs .speed/.stamina/.power/.guts/.wisdom attributes."""
    statvalues = [0, min(stats.speed, 901), min(stats.stamina, 901),
                  min(stats.power, 901), min(stats.guts, 901), min(stats.wisdom, 901)]
    total = sum((1 + math.floor(statvalues[stat] / 300.01)) * 0.05 for stat in course.course_set_status)
    return 1 + total / max(len(course.course_set_status), 1)


def get_course(course_id: int) -> CourseData:
    cached = _course_cache.get(course_id)
    if cached is not None:
        return cached
    raw = _raw()[str(course_id)]
    slopes = [Slope(s["start"], s["length"], s["slope"]) for s in raw.get("slopes", [])]
    if not is_sorted_by_start(slopes):
        slopes = sorted(slopes, key=lambda s: s.start)
    corners = [Corner(c["start"], c["length"]) for c in raw.get("corners", [])]
    straights = [Straight(s["start"], s["end"], s["frontType"]) for s in raw.get("straights", [])]
    course = CourseData(
        race_track_id=raw["raceTrackId"],
        distance=raw["distance"],
        distance_type=raw["distanceType"],
        surface=raw["surface"],
        turn=raw["turn"],
        course_set_status=list(raw.get("courseSetStatus", [])),
        corners=corners,
        straights=straights,
        slopes=slopes,
        # raw laneMax/10000 = max lane in course-widths (confirmed against
        # real track data: Tokyo turf, the doc's stated widest course, comes
        # out to exactly 1.5; narrow NAR dirt tracks come out around 1.1-1.2,
        # matching the doc's stated range).
        max_lane=raw.get("laneMax", 13000) / 10000.0,
    )
    _course_cache[course_id] = course
    return course

File: server/race_engine_rs/src/course_data.rs
//! Port of course_data.py.

use std::collections::HashMap;
use std::sync::OnceLock;

pub const SURFACE_TURF: i32 = 1;
pub const SURFACE_DIRT: i32 = 2;

#[derive(Clone, Copy, Debug)]
pub struct Corner {
    pub start: f64,
    pub length: f64,
}

#[derive(Clone, Copy, Debug)]
pub struct Straight {
    pub start: f64,
    pub end: f64,
    pub front_type: i32,
}

#[derive(Clone, Copy, Debug)]
pub struct Slope {
    pub start: f64,
    pub length: f64,
    pub slope: f64,
}

#[derive(Clone, Debug)]
pub struct CourseData {
    pub race_track_id: i32,
    pub distance: f64,
    pub distance_type: i32,
    pub surface: i32,
    pub turn: i32,
    pub course_set_status: Vec<i32>,
    pub corners: Vec<Corner>,
    pub straights: Vec<Straight>,
    pub slopes: Vec<Slope>,
    pub max_lane: f64,
}

pub fn is_sorted_by_start_f(starts: impl Iterator<Item = f64>) -> bool {
    let mut last = -1.0;
    for start in starts {
        if start <= last {
            return false;
        }
        last = start;
    }
    true
}

pub fn phase_start(distance: f64, phase: i32) -> f64 {
    match phase {
        0 => 0.0,
        1 => distance * 1.0 / 6.0,
        2 => distance * 2.0 / 3.0,
        3 => distance * 5.0 / 6.0,
        _ => panic!("invalid phase {phase}"),
    }
}

pub fn phase_end(distance: f64, phase: i32) -> f64 {
    match phase {
        0 => distance * 1.0 / 6.0,
        1 => distance * 2.0 / 3.0,
        2 => distance * 5.0 / 6.0,
        3 => distance,
        _ => panic!("invalid phase {phase}"),
    }
}

/// `speed`/`stamina`/`power`/`guts`/`wisdom` accessed positionally exactly
/// like the Python port's `[0, stats.speed, stats.stamina, ...]` array
/// (index 0 unused, ThresholdStat is 1-indexed in course_set_status).
pub fn course_speed_modifier(course: &CourseData, speed: f64, stamina: f64, power: f64, guts: f64, wisdom: f64) -> f64 {
    let statvalues = [0.0, speed.min(901.0), stamina.min(901.0), power.min(901.0), guts.min(901.0), wisdom.min(901.0)];
    let total: f64 = course
        .course_set_status
        .iter()
        .map(|&stat| (1.0 + (statvalues[stat as usize] / 300.01).floor()) * 0.05)
        .sum();
    1.0 + total / (course.course_set_status.len().max(1) as f64)
}

// -- data loading -------------------------------------------------------
// Populated by main.rs at startup from data/course_data.json (parsed with
// serde_json), via `init_courses`. Kept separate from this module's pure
// logic so the latter needed no JSON dependency while the linker was
// unavailable during development.

static COURSES: OnceLock<HashMap<i32, CourseData>> = OnceLock::new();

pub fn init_courses(courses: HashMap<i32, CourseData>) {
    COURSES.set(courses).ok();
}

pub fn get_course(course_id: i32) -> &'static CourseData {
    COURSES
        .get()
        .expect("init_courses must run before get_course")
        .get(&course_id)
        .unwrap_or_else(|| panic!("unknown course id {course_id}"))
}
